File: 2021/1/main.py
from typing import List


def read_inputs(fname: str) -> List[int]:
    with open(fname) as fp:
        data = fp.read()

    return [int(d) for d in data.split("\n")]


def simple_larger() -> int:
    inputs = read_inputs("./1/input/input")
    init = inputs[0]
    increased = 0

    for val in inputs:
        if val > init:
            increased += 1
        init = val

    return increased


def sliding_window_measurments() -> int:
    inputs = read_inputs("./1/input/input")
    increased = 0
    window_size = 3

    total_windows = len(inputs) - window_size

    window = 0
    previous_total = 0
    while window <= total_windows:
        current_total = sum(inputs[window: window + window_size])
        if window and current_total > previous_total:
            increased += 1

        previous_total = current_total
        window += 1

    return increased


if __name__ == "__main__":
    print(sliding_window_measurments())

File: 2021/1/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func readInput(fpath string) []int {
	var depths []int
	bData, err := os.ReadFile(fpath)
	check(err)

	for _, str := range strings.Split(string(bData), "\n") {
		depth, err := strconv.Atoi(str)
		check(err)

		depths = append(depths, depth)
	}

	return depths
}

func increasingDepth(depths *[]int) {
	increased := 0

	for i := range *depths {
		if i == 0 {
			continue
		}
		if (*depths)[i] > (*depths)[i-1] {
			increased++
		}
	}

	fmt.Println("No of times depth increased is: ", increased)
}

func sum(vals []int) int {
	result := 0
	for _, v := range vals {
		result += v
	}

	return result
}

func slidingWindowIncreasingDepth(depths *[]int) {
	increased := 0
	windowSize := 3
	previousMeasurment := 0
	currentMeasurment := 0

	for i := range (*depths)[:len(*depths)-2] {
		if i == 0 {
			previousMeasurment = sum((*depths)[i : i+windowSize])
			continue
		}
		currentMeasurment = sum((*depths)[i : i+windowSize])
		if currentMeasurment > previousMeasurment {
			increased++
		}
		previousMeasurment = currentMeasurment
	}

	fmt.Println("No of times depth increased in a sliding window calc is: ", increased)
}

func main() {
	depths := readInput("./input/input")
	increasingDepth(&depths)
	slidingWindowIncreasingDepth(&depths)
}
